Add a HTML5-based logview

brendandburns opened the pull request:
@itowlson 

This adds a webview based log viewer. It adds 4 capabilities:
* Filter matches
* Hide matches
* Show logs that follow match
* Show logs that preceed match

Here's some screenshots to show how it works...

No filtering:
![start](https://user-images.githubusercontent.com/5751682/51783007-dc799300-20e7-11e9-907c-612721a6eaf5.jpg)

Matches `/health`
![matches](https://user-images.githubusercontent.com/5751682/51783015-0632ba00-20e8-11e9-9127-19117dae83ac.jpg)

Hides `/health`
![not-matches](https://user-images.githubusercontent.com/5751682/51783017-1480d600-20e8-11e9-9a6b-00697a18943d.jpg)

Logs after a specific date:
![after](https://user-images.githubusercontent.com/5751682/51783019-1ba7e400-20e8-11e9-9cb4-82a477cd0a65.jpg)

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    private update() {
+        this.panel.title = `Kubernetes logs ${this.resource}`;
```
Logs with a capital L?  Perhaps also a separator?  To make it easy to scan I'd even suggest `Logs - ${this.resource}` so that less of it gets elided away when there are lots of tabs.

brendandburns replied:
done.

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+        <body>
+            <div style='position: absolute; top: 15px; left: 75%'>
+                <select id='mode' style='margin-bottom: 5px' onchange='eval()'>
+                    <option value='include'>Show Matches</option>
```
Capitalisation of options should be consistent - suggest sentence case (`Show matches` etc.).

brendandburns replied:
done.

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+        <body>
+            <div style='position: absolute; top: 15px; left: 75%'>
+                <select id='mode' style='margin-bottom: 5px' onchange='eval()'>
+                    <option value='include'>Show Matches</option>
```
I feel like the interaction with the regex box is not 100% clear.  I'd suggest 5 options:

* `Show all log entries`
* `Show log entries matching pattern`
* `Show log entries not matching pattern`
* `Show log entries up to first match`
* `Show log entries after first match`

(or you could use a label `Show log entries` with options `All`, `Matching pattern`, `Not matching pattern`, etc.)

brendandburns replied:
done.

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+        <body>
+            <div style='position: absolute; top: 15px; left: 75%'>
+                <select id='mode' style='margin-bottom: 5px' onchange='eval()'>
+                    <option value='include'>Show Matches</option>
```
Looking at your screenshots I worry this could overlap the first row of logs.  Another possible design would be toolbar style, with a bar at the top of the window reserved for these drop-downs, and the logs occupying the rest of the window a la:

```
+-------------------------------------------------+
| Show log entries: [...]  Match pattern: [...]   |
+-------------------------------------------------+
| 28.01.19 14:08:17 Printed l33t ASCII art        |
| ...                                             |
```

brendandburns replied:
done.

itowlson replied:
_shakes head at HTML_

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                    <option value='after'>Show logs after match</option>
+                    <option value='before'>Show logs before match</option>
+                </select>
+                <input type='text' id='regexp' onkeyup='eval()' placeholder='Regular Expression Filter' size='25'/>
```
I'd suggest just Filter or Match Pattern or something like that, and have the mention of regex in a tooltip or help button.  Though I guess if you move to the toolbar design you'll have more space to play with; but still I personally tend to prefer terser labels.

brendandburns replied:
done.

itowlson commented on src/components/kubectl/logs.ts:
```diff
@@ -86,7 +88,15 @@ function getLogsForContainer(
+
+    kubectl.invokeAsync(cmd).then(
+        (result: ShellResult) => {
+            panel.setInfo(result.stdout, resource);
```
Does this work with follow?  Surely `invokeAsync(...).then(...)` will show nothing until the shell invoke completes, and only _then_ send all logs (the whole of stdout) to the panel?  And `logs -f` will never complete without user intervention (which they can no longer do since there is no longer a terminal in play)?

brendandburns replied:
Nope, not currently. I need to write that code.

itowlson commented on src/components/kubectl/logs.ts:
```diff
@@ -86,7 +88,15 @@ function getLogsForContainer(
+
+    kubectl.invokeAsync(cmd).then(
+        (result: ShellResult) => {
+            panel.setInfo(result.stdout, resource);
```
Shouldn't this check for nonzero exit codes?

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                    <option value='after'>Show logs after match</option>
+                    <option value='before'>Show logs before match</option>
+                </select>
+                <input type='text' id='regexp' onkeyup='eval()' placeholder='Regular Expression Filter' size='25'/>
```
I'm a bit doubtful about `eval`-ing on key up - for one thing, it's quite possible for the regex to be in an invalid state at that point, and even if it is valid it could well be transient (the user in the middle of typing a pattern), so we could be filtering a lot of logs for nothing.  I'd suggest an `Apply Filter` button, or, if you want it to be automatic, a delay where it applies the filter once the user has not typed anything for half a second or a second or something like that.

brendandburns replied:
In practice it seems to work pretty well. I do want it to be automatic.

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                <div id='content' style="overflow-y: scroll; width: 100%; height: 100%"></div>
+            </div>
+            <script>
+              var orig = \`${this.content}\`.split('\\n');
```
Consider doing this in the LogsPanel rather than in the embedded page.  By having all this script in the HTML string, we lose any semblance of syntax checking.  In the cluster wizard I had buttons send messages to the host, which did all the fancy logic and then sent new content back to the webview, which kept the on-page script fairly simple.

itowlson replied:
Yes, okay

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                // TODO: This could really be improved to avoid the double loop and list construction.
+                if (regexp.length > 0) {
+                    var mode = $('#mode').val();
+                    var regex = new RegExp(regexp);
```
Do we need error handling around this in case the user input is invalid regex text?

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                            break;
+                        case 'before':
+                            content = [];
+                            for (var i = 0; i < orig.length; i++) {
```
`for (const line of orig)`?

itowlson replied:
Or use `findIndex` to find the index of the first matching line and then slice, though this will require no-match handling.

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                            break;
+                        case 'after':
+                            var i = 0;
+                            for (; i < orig.length; i++) {
```
Would it not be easier to use `findIndex`?

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                } else {
+                    content = orig;
+                }
+                $('#content').html('<code><pre>' + content.join('\\n') + '</pre></code>');
```
Do you need both `code` and `pre`?

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+                    var mode = $('#mode').val();
+                    var regex = new RegExp(regexp);
+                    switch (mode) {
+                        case 'include':
```
Does this pay any regard to multiline log entries e.g. if I have:

```
2019/01/28 14:37:00 Pod failed to start because:
* There was a fish in the percolator
* The owls were not what they seemed
* Fire walked with me
2019/01/28 14:37:02 Container restarted
```

and match `failed to start`, will I get the essential owl-based diagnostics or will I need to e.g. note the timestamp, remove the filter and scroll?

itowlson replied:
Huh, 'go to timestamp' would be a really neat command for this.

itowlson replied:
As would 'go to first/next match' so you could easily see the matched log in its context.

Perhaps we should have a command to open the logs in a normal editor window, so people could use all the fancy find features built into VS Code and into extensions (I saw there was at least one line filtering extension though I have no idea how well it works!).  Though that would still struggle with the Follow scenario.

brendandburns replied:
Future features (but I agree...)

itowlson commented on src/components/kubectl/logs.ts:
```diff
@@ -86,7 +88,15 @@ function getLogsForContainer(
 
-    kubectl.invokeInSharedTerminal(cmd);
+    const resource = `${podSummary.namespace}/${podSummary.name}`;
+    const panel = LogsPanel.createOrShow('TODO', 'Loading \u25CC', resource);
```
TODO?

brendandburns replied:
removed.

itowlson commented on src/components/kubectl/logs.ts:
```diff
@@ -86,7 +88,15 @@ function getLogsForContainer(
 
-    kubectl.invokeInSharedTerminal(cmd);
+    const resource = `${podSummary.namespace}/${podSummary.name}`;
+    const panel = LogsPanel.createOrShow('TODO', 'Loading \u25CC', resource);
```
Consider `...` rather than the 'dotted circle' character?

brendandburns replied:
done.

itowlson commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+import * as path from 'path';
```
Is `path` used?

brendandburns replied:
removed.

squillace commented:
@itowlson I'm kinda with Brendan here, and that despite my desire to be against him. Why on earth is it a Kantian imperative to do all that extra stuff? I'm thrilled that bits and pieces of very useful stuff will appear that helps the overall devux. But doing the full operator is still a fairly hard boundary, right?

itowlson commented:
After a bit more nosing around, I don't think the extensions I thought might make working with logs as documents actually do what I hoped.  So maybe handling the display ourselves is indeed the way to go.

One thing I would suggest as a bridge is an "Open in Editor" command (which could go in the toolbar region if you adopt that suggestion), which simply dumps the log text into an editor window from where the user can save it, search with built-in facilities, delete stuff that is not of interest, etc.  That gives users a quick way to get over into editorland if that's what they want, while still putting the custom experience first.

Finally I still feel we need to support the 'Follow Logs' scenario, unless I'm misreading this PR and that's still going to the terminal.  Which would be fine - I'm happy to take this transition in steps.

brendandburns commented:
@itowlson thanks for the feedback. I'll add the "open in document" option and yes, follow logs is absolutely next on the list, I just wanted to get agreement on the added functionality before I did the work.

I'll address your detailed comments, and see how hard it is to add follow logs to this PR. It might be easier to do it in a follow on.

Thanks again for the thoughts and the feedback!

brendandburns commented on src/components/logs/logsWebview.ts:
```diff
@@ -0,0 +1,145 @@
+    private content: string;
+    private resource: string;
+
+    public static createOrShow(extensionPath: string, content: string, resource: string): LogsPanel {
```
done.

brendandburns commented:
Comments addressed, new screen shot is attached.
![logview](https://user-images.githubusercontent.com/5751682/52160429-0bed4a00-266a-11e9-9bfa-45fc04cedc44.png)

itowlson commented:
You've got a few CI errors due to a bunch of strictness options being turned on in `master`.  The CI logs don't seem to actually _list_ the warnings for some reason, but when I run `tsc -p .` locally I get:

```
src/components/describe/describeWebview.ts(2,1): error TS6133: 'Dictionary' is declared but its value is never read.
src/components/kubectl/logs.ts(9,1): error TS6133: 'host' is declared but its value is never read.
src/components/kubectl/logs.ts(11,1): error TS6133: 'ShellResult' is declared but its value is never read.
src/components/kubectl/logs.ts(96,13): error TS2532: Object is possibly 'undefined'.
src/components/kubectl/logs.ts(97,67): error TS2532: Object is possibly 'undefined'.
src/components/kubectl/logs.ts(99,27): error TS2532: Object is possibly 'undefined'.
src/components/webpanel/webpanel.ts(12,30): error TS7017: Element implicitly has an 'any' type because type 'Map<string, T>' has no index signature.
src/components/webpanel/webpanel.ts(17,13): error TS2322: Type 'undefined' is not assignable to type 'T'.
src/components/webpanel/webpanel.ts(27,9): error TS7017: Element implicitly has an 'any' type because type 'Map<string, T>' has no index signature.
src/components/webpanel/webpanel.ts(43,42): error TS6133: 'e' is declared but its value is never read.
src/components/webpanel/webpanel.ts(57,16): error TS7017: Element implicitly has an 'any' type because type 'Map<string, T>' has no index signature.
src/components/webpanel/webpanel.ts(69,24): error TS7010: 'update', which lacks return-type annotation, implicitly has an 'any' return type.
```

None of these will affect behaviour; I can fix them up before merge if you want.

itowlson commented:
I'm playing with the filter stuff and I've noticed some performance hiccups with large volumes of logs.  I created a `logspouter` pod for testing and it has perhaps been spouting too often for too long now... but the result is that when I switch from `Show log entries: all` to `that match`, it locks up for several seconds (the drop-down stays open and I can't click on the filter box).  Then when I type the first character of the filter, it freezes again.  Only as I type enough characters to filter the logs down to a more reasonable amount does it become responsive.

`logspouter` may well be unrepresentative of real k8s pod logs, as it was specifically created to emit logs often enough that I could experiment with tailing ideas - but then again a real pod might have lived _far_ longer than logspouter's 5 days.  You have more experience than me of the sort of volumes that are realistic, and if you feel that it's responsive enough with realistic data sets then I trust you!  But it's definitely slower than I would like in this particular test case.

(I can give you access to `logspouter` if you want to try it - just let me know.)

itowlson commented:
Okay sorry to keep leaving naggy comments but there is also a glitch if you switch away from the logs window.  When you switch back it appears _completely blank_ (no controls or `Loading...` message).  If, as with the `logspouter` test, it takes a long time to render, then there is no feedback during this time.

There is an option `retainContextWhenHidden` which appears to fix this - it does incur memory overhead, which is not exactly something VS Code needs more of, but worth it I feel....?  But again maybe it isn't going to be an issue for more realistic data sets.

itowlson commented:
If it helps to provide context, at the time I was playing with this stuff, `logspouter` had emitted about 186000 lines of logs with a total size of about 17MB.

itowlson commented:
Re the toolbar scrolling, I was playing with this for something else and this worked for me: https://stackoverflow.com/questions/9677894/header-div-stays-at-top-vertical-scrolling-div-below-with-scrollbar-only-attach.  The hardwired height is a bit of a cheat but we're not exactly writing a HTML style tutorial here... *grin*

brendandburns commented:
Thanks, yeah, I had the toolbar pinned before, and I agree, I'll figure out to pin it again.

And thanks for the performance reports, I'll try to repro and fix.

brendandburns commented:
Ok, so I put it under the electron performance microscope :) and it turns out that nearly all of the time is spent _rendering_, filtering is pretty fast, it's the actual pixel spitting that is slow.

I cleaned this up somewhat. Here's what I did:

* Moved the filtering to an async method, so the UX won't freeze
* Split the rendering up into batches via `setTimeout` so that the UX won't freeze and it will feel responsive (this actually works pretty slick, the only downside is that the scrollbar wiggles a little, but it's barely noticeable)
* Added [`retainContentWhenHidden`](https://code.visualstudio.com/api/extension-guides/webview#retaincontextwhenhidden) to the Webview so that the switching tabs thing doesn't happen.

At this point, I think it feels pretty good in terms of responsiveness.

Please take another look. I'll start a new PR to add support for following logs.

itowlson commented:
Performance is great now!  Thanks!

Unfortunately... (and apologies if this is something you're still working on)... the toolbar is now _pinned_ but the content scrolls underneath it.  The logs need to be in a separate scrolling `div`.

![image](https://user-images.githubusercontent.com/865538/52539037-88a4c480-2dde-11e9-85ae-d143ad09e6c8.png)

brendandburns commented:
Fixed. (though the length of time it took me to figure out how to do it is a clear sign I shouldn't be slinging CSS)

Please take another look...

itowlson commented:
Urgh, I am sorry to be a pain but... now the docking is working but the scroll bar is not visible...

![image](https://user-images.githubusercontent.com/865538/52745896-1da8f700-3045-11e9-858d-9a22731acfaf.png)

...unless I horizontally scroll to it, which cuts off the beginning of the log lines...

![image](https://user-images.githubusercontent.com/865538/52745807-e5a1b400-3044-11e9-91fd-d586f746857b.png)

The lines in the log file are not so long that they should require horizontal scrolling, and even if they were, it would be good if the vertical scroll bar were always visible...

(I have artifically ensmallened the VS Code window for screenshot purposes... if I embiggen it then I get a _different_ misbehaviour:

![image](https://user-images.githubusercontent.com/865538/52746092-9e67f300-3045-11e9-806d-5d5a1882a3ca.png)

Here the scroll bar does not dock to the RHS of the window.  So it does feel like a div width issue.)

itowlson commented:
I did have a go at fixing the scrollbar positioning but have had no luck I'm afraid - I think it's tricky with the progressive loading...

brendandburns commented:
Yeah, I noticed that issue and hoped no one else would *sigh*

I'll work on fixing things some other way...

brendandburns commented:
@itowlson a miracle occurred and I was enlighted about CSS. Well, not actually, but I did figure out what was needed here (I think) please re-check!

Thanks

itowlson commented:
I understand miracles are a core element of the CSS design workflow.  Thanks, will take a look!

brendandburns commented:
Thanks (insert Neo/Matrix "I know CSS" meme here :)

I'll work on the follow logs implementation next.
